File: CPP/ByTier/D/2112_D_Reachability_and_Tree.cpp
// Problem: CF 2112 D - Reachability and Tree
// https://codeforces.com/contest/2112/problem/D

/*
D. Reachability and Tree
time limit per test2 seconds
memory limit per test256 megabytes

Task:
Given an undirected tree with n vertices, determine whether it is possible to assign a direction to each edge so that exactly n pairs (u,v) exist such that there is a directed path from u to v.
If possible, output any valid orientation of edges.

Algorithms/Techniques:
- Graph traversal (DFS)
- Tree properties and edge orientation logic
- Identifying special structure in tree for valid assignment

Time Complexity: O(n) per test case
Space Complexity: O(n) per test case

The approach is based on the idea that for a tree to have exactly n good pairs,
we must orient edges such that only vertices with degree 2 can be part of a cycle-like structure
that allows for exactly n directed pairs. A key insight is that if there exists a vertex of degree 2,
we can use a DFS-like traversal strategy where we direct edges along specific paths to achieve n good pairs.

For each test case:
- If there's a node with degree 2, construct the solution by traversing from neighbors of this node
- The solution involves directing edges away from the central vertex of degree 2,
  forming a tree-like structure that allows for exactly n good pairs.
*/

#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;
typedef long long ll;
vector<vector<int>> gr;

// Function f(x, p, c): recursive DFS to orient edges
// x: current node, p: parent node, c: flag indicating direction of edge (0 or 1)
void f(int x, int p, int c) {
  for (auto y : gr[x]) {             // Iterate through neighbors of current node
    if (c)                           // If edge should be directed from x to y
      cout << x + 1 << ' ' << y + 1 << '\n';
    if (y != p)                      // Avoid going back to parent
      f(y, x, c ^ 1);                // Recurse with flipped direction flag
  }
}

void solve() {
  int n;
  cin >> n;
  gr.clear();
  gr.resize(n);
  for (int i = 0; i < n - 1; ++i) {  // Read the tree edges and build adjacency list
    int u, v;
    cin >> u >> v;
    u--, v--;
    gr[u].push_back(v);
    gr[v].push_back(u);
  }
  
  for (int i = 0; i < n; ++i) {      // Find a node with degree 2
    if (gr[i].size() == 2) {
      cout << "YES\n";
      f(gr[i][0], i, 1);           // Start traversing from one neighbor
      f(gr[i][1], i, 0);           // Continue from another neighbor
      cout << i + 1 << ' ' << gr[i][1] + 1 << '\n'; // Final edge direction
      return;
    }
  }
  
  cout << "NO\n";                    // No valid orientation found
}

int main() {
  ios_base::sync_with_stdio(0);
  cin.tie(0);
  int t;
  cin >> t;
  while (t--)
    solve();
}


// https://github.com/VaHiX/codeForces/

File: Java/ByRound/2112/2112_D_Reachability_and_Tree.java
// Problem: CF 2112 D - Reachability and Tree
// https://codeforces.com/contest/2112/problem/D

/*
 * Problem: D. Reachability and Tree
 *
 * Purpose:
 *   Given an undirected tree with n vertices, determine if we can assign directions to edges such that
 *   exactly n ordered pairs (u,v) exist where u can reach v via directed paths.
 *
 * Approach:
 *   1. Build the adjacency list representation of the tree.
 *   2. Perform BFS from node 1 to compute in-degrees and out-degrees for each node, alternating
 *      directions based on BFS level.
 *   3. Find a node with total degree 2 (i.e., in-degree + out-degree = 2). If none exists or n=2,
 *      return "NO".
 *   4. Do another BFS starting from such a node to orient the edges so that each edge is oriented
 *      toward the next BFS level, ensuring exactly n good pairs.
 *
 * Time Complexity: O(n) per test case, since we process each node and edge once in BFS traversals.
 * Space Complexity: O(n) for storing adjacency list, visited array, and auxiliary data structures.
 */

import java.io.*;
import java.util.*;

public class Main {
    public static void fx(BufferedReader br, StringBuilder out) throws IOException {
        int n = Integer.parseInt(br.readLine());
        List<List<Integer>> ls = new ArrayList<>();
        for (int i = 0; i <= n; i++) ls.add(new ArrayList<>());
        for (int i = 0; i < n - 1; i++) {
            String[] parts = br.readLine().split(" ");
            int x = Integer.parseInt(parts[0]);
            int y = Integer.parseInt(parts[1]);
            ls.get(x).add(y);
            ls.get(y).add(x);
        }
        int[] inDeg = new int[n + 1];
        int[] outDeg = new int[n + 1];
        boolean[] vis = new boolean[n + 1];
        Queue<Integer> q = new ArrayDeque<>();
        q.offer(1);
        vis[1] = true;
        int k = 1;
        while (!q.isEmpty()) {
            int t = q.size();
            for (int i = 0; i < t; i++) {
                int cur = q.poll();
                int deg = 0;
                for (int j : ls.get(cur)) {
                    if (!vis[j]) {
                        q.offer(j);
                        vis[j] = true;
                    }
                    deg++;
                }
                // Alternate direction assignment: even levels get out-degree, odd get in-degree
                if (k == 1) outDeg[cur] = deg;
                else inDeg[cur] = deg;
            }
            k = (k + 1) % 2;
        }
        int ind = -1;
        // Find any node with degree sum of 2, which will be a bridge connecting two components
        for (int i = 1; i <= n; i++) {
            if (inDeg[i] + outDeg[i] == 2) {
                ind = i;
                break;
            }
        }
        // If no such node found or only two nodes, impossible to achieve exactly n good pairs
        if (ind == -1 || n == 2) {
            out.append("NO\n");
            return;
        }
        List<int[]> res = new ArrayList<>();
        q.clear();
        vis = new boolean[n + 1];
        vis[ind] = true;
        k = 0;
        // Process all neighbors of 'ind' first
        for (int i : ls.get(ind)) {
            q.offer(i * 2 + k); 
            vis[i] = true;
            if (k == 1) res.add(new int[]{i, ind});
            else res.add(new int[]{ind, i});
            k = 1;
        }
        while (!q.isEmpty()) {
            int val = q.poll();
            int x = val / 2;
            int tk = val % 2;
            for (int i : ls.get(x)) {
                if (!vis[i]) {
                    q.offer(i * 2 + (tk ^ 1));
                    // Set direction based on whether we came from parent or child
                    if (tk == 0) res.add(new int[]{i, x});
                    else res.add(new int[]{x, i});
                    vis[i] = true;
                }
            }
        }
        out.append("YES\n");
        for (int[] r : res) {
            out.append(r[0]).append(" ").append(r[1]).append("\n");
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder out = new StringBuilder();
        int t = Integer.parseInt(br.readLine());
        while (t-- > 0) {
            fx(br, out);
        }
        System.out.print(out);
    }
}


// https://github.com/VaHiX/CodeForces/